dotnet-authserver/src/TeacherIdentity.AuthServer/Journeys/SignInJourney.cs

```diff
@@ -0,0 +1,141 @@
+using Microsoft.AspNetCore.Mvc;
+using TeacherIdentity.AuthServer.Models;
+
+namespace TeacherIdentity.AuthServer.Journeys;
+
+public abstract class SignInJourney
+{
+    protected SignInJourney(
+        AuthenticationState authenticationState,
+        HttpContext httpContext,
+        IdentityLinkGenerator linkGenerator)
+    {
+        AuthenticationState = authenticationState;
+        HttpContext = httpContext;
+        LinkGenerator = linkGenerator;
+    }
+
+    public AuthenticationState AuthenticationState { get; }
+
+    public HttpContext HttpContext { get; }
+
+    public IdentityLinkGenerator LinkGenerator { get; }
+
+    public abstract bool IsFinished();
+
+    public virtual string GetStartStep() => Steps.Email;
+
+    public virtual string GetStepUrl(string step) => step switch
```

Comment:
The `GetStepUrl`, `GetNextStep`, `GetPreviousStep` and `CanAccessStep` methods are expected to overriden by deriving types. The idea is that deriving types implement the logic for the pages their journey-specific pages then delegate to the `base` implementation for the shared pages (e.g. `Email`, `EmailConfirmation`).

Another approach would be to make these all abstract and have each deriving type implement the method completely.

Reply:
That makes sense. In reality the base class can't completely encapsulate the `Email` + `EmailConfirmation` steps any way because it won't know what comes next.